File: baekjoon_1717/1717.kt
import java.io.BufferedReader
import java.io.InputStreamReader

fun find(set: MutableList<Int>, value: Int): Int {
    if (value == set.elementAt(value)) return value

    set[value] = find(set, set[value])
    return set[value]
}

fun union(set: MutableList<Int>, a: Int, b: Int) {
    val x = find(set, a)
    val y = find(set, b)

    set[y] = x
}

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = System.out.bufferedWriter()

    val (n, m) = br.readLine().split(" ").map { it.toInt() }
    val parent = (0..n+1).toMutableList()
    repeat(m) {
        val input = br.readLine().split(" ").map { it.toInt() }
        val cmd = input[0]
        val (a, b) = input[1] to input[2]
        when (cmd) {
            0 -> {
                union(parent, a, b)
            }
            1 -> {
                if (find(parent, a) == find(parent, b))
                    bw.write("YES\n")
                else
                    bw.write("NO\n")
            }
        }
    }

    br.close()
    bw.close()
}

File: baekjoon_1717/1717.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "strconv"
    "strings"
)

func find(set []int, value int) int {
    if value == set[value] {
        return value
    }
    set[value] = find(set, set[value])
    return set[value]
}

func union(set []int, a, b int) {
    x := find(set, a)
    y := find(set, b)
    if x != y {
        set[y] = x
    }
}

func main() {
    reader := bufio.NewReader(os.Stdin)
    writer := bufio.NewWriter(os.Stdout)
    defer writer.Flush()

    input, _ := reader.ReadString('\n')
    nums := strings.Fields(input)
    n, _ := strconv.Atoi(nums[0])
    m, _ := strconv.Atoi(nums[1])

    parent := make([]int, n+1)
    for i := 0; i <= n; i++ {
        parent[i] = i
    }

    for i := 0; i < m; i++ {
        input, _ := reader.ReadString('\n')
        commands := strings.Fields(input)
        cmd, _ := strconv.Atoi(commands[0])
        a, _ := strconv.Atoi(commands[1])
        b, _ := strconv.Atoi(commands[2])

        if cmd == 0 {
            union(parent, a, b)
        } else if cmd == 1 {
            if find(parent, a) == find(parent, b) {
                writer.WriteString("YES\n")
            } else {
                writer.WriteString("NO\n")
            }
        }
    }
}
